feat: add optional token and secret-envs

NikitaCOEUR opened the pull request:
This pull request includes significant updates to the GitHub Actions workflows and related documentation to support the use of GitHub Application tokens. The most important changes include adding new inputs and secrets for GitHub Application tokens, updating workflow steps to handle these tokens, and modifying related action files to support the new functionality.

### Workflow Enhancements:

* [`.github/workflows/docker-build-images.md`](diffhunk://#diff-c5f28bd5dc1e2fc23df850e9ced6d77d54b09b297618a1f4915e6fd63814a439R82-R93): Added new inputs `app-id` and `app-token-to-docker-secret-envs-keys`, and a new secret `app-key` to support GitHub Application tokens. [[1]](diffhunk://#diff-c5f28bd5dc1e2fc23df850e9ced6d77d54b09b297618a1f4915e6fd63814a439R82-R93) [[2]](diffhunk://#diff-c5f28bd5dc1e2fc23df850e9ced6d77d54b09b297618a1f4915e6fd63814a439R104-R105) [[3]](diffhunk://#diff-c5f28bd5dc1e2fc23df850e9ced6d77d54b09b297618a1f4915e6fd63814a439R119-R120)
* [`.github/workflows/docker-build-images.yml`](diffhunk://#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR79-R96): Updated the workflow to include steps for creating a GitHub Application token, generating secrets, and using these tokens in the build process. [[1]](diffhunk://#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR79-R96) [[2]](diffhunk://#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR108-R111) [[3]](diffhunk://#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR307-R314) [[4]](diffhunk://#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR323-R345) [[5]](diffhunk://#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR361-R362) [[6]](diffhunk://#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR425-R432) [[7]](diffhunk://#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR441-R442)

### Action File Updates:

* [`actions/docker/build-image/README.md`](diffhunk://#diff-d384d3b448dd1ffc97a4d317d15a12b3360d3f8731aece711f29b2d0fee5a8f3R102-R110): Added descriptions for new inputs `secret-envs` and `token` for exposing secrets and using custom tokens. [[1]](diffhunk://#diff-d384d3b448dd1ffc97a4d317d15a12b3360d3f8731aece711f29b2d0fee5a8f3R102-R110) [[2]](diffhunk://#diff-d384d3b448dd1ffc97a4d317d15a12b3360d3f8731aece711f29b2d0fee5a8f3R130-R131)
* [`actions/docker/build-image/action.yml`](diffhunk://#diff-b1a31054574d361286a45d0269d77a0c0845e02d0ff84821308827c5e370a869R101-R109): Added new inputs `secret-envs` and `token`, and updated steps to use these inputs. [[1]](diffhunk://#diff-b1a31054574d361286a45d0269d77a0c0845e02d0ff84821308827c5e370a869R101-R109) [[2]](diffhunk://#diff-b1a31054574d361286a45d0269d77a0c0845e02d0ff84821308827c5e370a869R127) [[3]](diffhunk://#diff-b1a31054574d361286a45d0269d77a0c0845e02d0ff84821308827c5e370a869L191-R204) [[4]](diffhunk://#diff-b1a31054574d361286a45d0269d77a0c0845e02d0ff84821308827c5e370a869R239)

### Metadata Action Updates:

* [`actions/docker/get-image-metadata/README.md`](diffhunk://#diff-6b3003a113b8d0aa4f768e63f20ae0ccab2b72891de5e67bbd806eb95c8ce62bR48-R51): Added a new input `token` for using a custom token for repository checkout. [[1]](diffhunk://#diff-6b3003a113b8d0aa4f768e63f20ae0ccab2b72891de5e67bbd806eb95c8ce62bR48-R51) [[2]](diffhunk://#diff-6b3003a113b8d0aa4f768e63f20ae0ccab2b72891de5e67bbd806eb95c8ce62bR63)
* [`actions/docker/get-image-metadata/action.yml`](diffhunk://#diff-ea046ec82ea2211b4a71615614f13cd10afb67a9b38fa9c8558dc8a7b0e3b278R38-R40): Added a new input `token`, and updated steps to use this token. [[1]](diffhunk://#diff-ea046ec82ea2211b4a71615614f13cd10afb67a9b38fa9c8558dc8a7b0e3b278R38-R40) [[2]](diffhunk://#diff-ea046ec82ea2211b4a71615614f13cd10afb67a9b38fa9c8558dc8a7b0e3b278L65-R70) [[3]](diffhunk://#diff-ea046ec82ea2211b4a71615614f13cd10afb67a9b38fa9c8558dc8a7b0e3b278R116)

These changes enhance the security and flexibility of the workflows by allowing the use of GitHub Application tokens, which can provide more granular permissions and better security practices.

### How to use it

```
  build-ci:
    # FIXME: This is a workaround for having workflow actions. See https://github.com/orgs/community/discussions/38659
    permissions:
      id-token: write
      contents: read
      packages: write
      issues: read
      pull-requests: read
      actions: write
    uses: Groupe-Hevea/ci-github-container/.github/workflows/docker-build-images.yml@feat/add-github-application-token-option # hoverkraft-tech/ci-github-container/.github/workflows/docker-build-images.yml@5b53f22d16042c9ddf944be7323d47e2b977310d # 0.20.2
    needs: prepare-variables-ci
    with:
      oci-registry: ${{ vars.OCI_REGISTRY }}
      images: |
        [
          {
            "name": "ci",
            "context": ".",
            "dockerfile": "Dockerfile",
            "build-args": { "PROD_MODE": "" },
            "target": "ci",
            "platforms": ["linux/amd64"]
          }
        ]
      app-id: ${{ vars.APP_ID }}
      app-token-to-docker-secret-envs-keys: "GITHUB_TOKEN"
    secrets:
      oci-registry-password: ${{ secrets.OCI_REGISTRY_PASSWORD }}
      app-key: ${{ secrets.APP_KEY }}

```

It will generate an installation token based on your installed Github Application and map secret named `GITHUB_TOKEN` to env var `APP_TOKEN` with generated token via `actions/create-github-app-token`.

> [!NOTE]  
> We can NOT use application token instead of `oci-registry-password` for the moment, https://docs.github.com/en/packages/learn-github-packages/about-permissions-for-github-packages#about-scopes-and-permissions-for-package-registries
![{BBFC8EEE-1CD5-4357-BAB5-1564F8AA1FB6}](https://github.com/user-attachments/assets/d62839ff-0b1e-49b3-88c4-c8d0df9cd2f3)


> [!WARNING]  
> Depends on https://github.com/hoverkraft-tech/ci-github-common/pull/298 for `checkout ci-github-common/actions/checkout` a commit dedicated to switch to the new version is needed before merging this PR.

NikitaCOEUR commented:
We need this because we're starting to get errors due to the rate limit imposed on PATs.

neilime commented:
Hi @NikitaCOEUR, can you rebase it (main is using the new version of common actions)

FYI README files are auto updated regarding related action.yml files so you dont have to modify them.

Thanks

NikitaCOEUR commented:
> Hi @NikitaCOEUR, can you rebase it (main is using the new version of common actions)
> 
> FYI README files are auto updated regarding related action.yml files so you dont have to modify them.
> 
> Thanks

Hi @neilime, 
I've just rebased without the README changes..

NikitaCOEUR commented:
> Hi @NikitaCOEUR, from what I understand, this PR allows to pass github app token for 2 use cases:
> 
>     * checkout
> 
>     * docker build secret
> 
> 
> Can you explain me why you need to pass this custom token (instead of default GITHUB_TOKEN)?
> 

GITHUB_TOKEN does not have the right to interact with other private repositories in our organization.
We use a PAT-type token that we add to our composer's configuration to download internal vendors. If we use the classic GITHUB_TOKEN, we have to deal with rights issues.
We don't have the same rights problems with the PAT, but it is restricted to 5000 api calls per hour. We sometimes get stopped by rate limits when building our applications.
That's why I'd like to use a GITHUB App token instead, to increase the rate limit to 15,000 calls/hour.

However, we can't use the application token generated by one job in another job (this is restricted to within the same job). This is why I wanted to enable the generation of the application token at the workflow level and map it to an ENV secret so that it could be reused in the same build.

As far as checkouts are concerned, if the GITHUB_TOKEN allows us to interact with all the necessary repo, there's no need to use the token generated by the github app.

neilime commented:
@NikitaCOEUR The build images workflow is used in several private organizations without any issues.

Are you sure the permissions are well configured:

https://github.com/hoverkraft-tech/ci-github-container/blob/main/.github/workflows/docker-build-images.md

NikitaCOEUR commented:
I think the problem we're having stems from the fact that we're using a PAT which we inject into our built images and which uses this token to download all the vendors, we're obliged to use a PAT given that some of the vendors are in our private repositories.

Perhaps setting these repositories to "internal" would allow us to use GITHUB_TOKEN to download these vendors.

The other solution is to use a Github APP token with the application configured to interact with our private repositories.

Once again, we don't have a rights problem, only a rate-limit problem!

neilime commented:
Ok so the issue is only for build arg token that should be an app token instead of PAT.

No need for checking out, right?

NikitaCOEUR commented:
I'd decided to run everything through the Github App, but on reflection, I don't think it's necessary for checking out. You're right.

neilime commented:
@NikitaCOEUR can you try main branch it is supposed to match with your needs:

Inputs `build-secret-github-app-token-env` & `build-secret-github-app-id` 
Secret `build-secret-github-app-key`

https://github.com/hoverkraft-tech/ci-github-container/blob/main/.github/workflows/docker-build-images.md

I guess we can close this PR.

If you have any feedback, feel free to open an issue or a PR.

Thanks

NikitaCOEUR commented:
Unfortunately, it doesn't work, because of the way the token is created here : https://github.com/hoverkraft-tech/ci-github-container/blob/main/.github/workflows/docker-build-images.yml#L345-L351

I created it like this: 
https://github.com/hoverkraft-tech/ci-github-container/pull/361/files#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR314

I specified the organization in order to generate a token that could access all the repositories to which the github application was authorized to access in our organization. By not specifying the organization, the generated token has restricted access to the current repository only : 
```
Run actions/create-github-app-token@v1
owner and repositories not set, creating token for the current repository 
```

It may be necessary to be able to specify certain parameters when generating the token, or to decide to add the organization directly. Otherwise, this token is no more interesting than the classic GITHUB_TOKEN.

neilime commented:
@NikitaCOEUR can you try last main versio? `build-secret-github-app-owner` input has been added

NikitaCOEUR commented:
Um, it's better regarding the creation of the token : 
![{EFDE8A53-EE35-4CBD-B388-BD6A0236855D}](https://github.com/user-attachments/assets/03bd1c2d-80c8-4a96-9e1e-1691de49ba61)

On the other hand, there seems to be a problem with `build-secret-github-app-token-env`, in fact, I need to map the token generated by the github app to GITHUB_TOKEN, it seems to work, on the other hand `secret-envs` requires EVERY secret specified to be mapped to an existing environment variable. Like here : 
https://github.com/hoverkraft-tech/ci-github-container/pull/361/files#diff-54a2aad43325eed5cae855e7a691fd0b09eb64ffdb2e2c4dd855672b69605b5cR344

I think you forgot to publish the `GITHUB_APP_TOKEN` environment variable so that the `secret-envs` that refer to it can access it.

neilime commented:
@NikitaCOEUR the export of GITHUB_APP_TOKEN env var is done there: https://github.com/hoverkraft-tech/ci-github-container/blob/main/.github/workflows/docker-build-images.yml#L376

NikitaCOEUR commented:
Yes, you're right, we can see the good inputs: 
![image](https://github.com/user-attachments/assets/adfc61b9-ebd4-4afc-88e4-c591a4fa61f7)

However, something surprising is that secret-envs do not appear in the command line that is executed during the build action : 
![image](https://github.com/user-attachments/assets/7f5462e8-0049-4b37-83e9-aa0dcb52fb9c)

The same command line when it worked with my PR, you can see that there is MUCH more information, I have the impression that there is a problem with the format of the input `build-secret-github-app-token-env` which makes it truncate the command line ....
![image](https://github.com/user-attachments/assets/6040780d-0d69-4c56-b500-17bb6af50b79)

NikitaCOEUR commented:
But, for you, it seems ok: 
![image](https://github.com/user-attachments/assets/919d4350-850a-4670-81c9-f41b2e7e69c0)

NikitaCOEUR commented:
I seem to have a line break at the beginning: 
![{525576CC-2031-443C-803E-10E6994F1207}](https://github.com/user-attachments/assets/6d91ee74-74d7-4e73-a7cb-4eed8d54e145)
whereas for you: 
![{372F4D68-C54F-418A-BB56-C6C124DA643C}](https://github.com/user-attachments/assets/2841535b-567c-42ca-96c0-e1da60f54da3)

NikitaCOEUR commented:
I set the `build-secret-github-app-token-env` entry like this:
![{16154231-5841-4166-9267-790874CDABF9}](https://github.com/user-attachments/assets/1de1371e-97a6-456a-87e8-0a2ef9df0d1b)
And you : 
![{7E87ECB4-A995-4111-B7E6-867BFBE42111}](https://github.com/user-attachments/assets/a1163686-4030-4317-bb6e-bc29f7bdb34c)
So, the same way; but we can see you have other secrets when calling the action : 
![{F543B357-4E5E-4976-929A-0FA152E78637}](https://github.com/user-attachments/assets/e8afb7bf-599c-4b17-b86f-6b54f1c45e39)
it seems to come from here : https://github.com/hoverkraft-tech/ci-github-container/blob/a381f650910cc0f08fd07945ca6354381d753c6f/.github/workflows/__test-workflow-docker-build-images.yml#L70-L73

So I think there's a problem with this step: 
https://github.com/hoverkraft-tech/ci-github-container/blob/a381f650910cc0f08fd07945ca6354381d753c6f/.github/workflows/docker-build-images.yml#L383-L386

NikitaCOEUR commented:
@neilime 
It should be : 
```
if (secretEnvs) {
  secretEnvs += '\n';
}
secretEnvs += buildSecretGithubAppTokenEnv 
               .split('\n') 
               .filter(Boolean) 
               .map(key => `${key}=${githubAppTokenEnvName}`).join('\n');             
```
Instead of : 
```
secretEnvs += '\n' + buildSecretGithubAppTokenEnv 
               .split('\n') 
               .filter(Boolean) 
               .map(key => `${key}=${githubAppTokenEnvName}`).join('\n');             
```

Or something like that

neilime commented:
@NikitaCOEUR it should be good now

NikitaCOEUR commented:
It still doesn't work, I have the impression that the format doesn't suit it. I remember I had the same problem and had to encircle my secret-envs with quotation marks. 
Here, it doesn't work : 
![{7E5AFFB9-CC2D-43DA-844C-3E9F9EDF9B10}](https://github.com/user-attachments/assets/0e695508-1e6c-4ab4-8cc9-90e3af5c70ee)
Here, it worked : 
![{7FA0F122-D330-45AA-ABBD-77905A4CE36D}](https://github.com/user-attachments/assets/64e21864-3380-410a-bbcc-c69f58a8ee7d)

But it works well in your tests :
![{982DBA4A-F092-4FDF-B4E7-BAFE621F27F5}](https://github.com/user-attachments/assets/549165a7-0ffd-463e-99f3-955bbb4dad06)

I don't understand why the order is truncated. : 
![image](https://github.com/user-attachments/assets/c1b99c58-843c-4d2d-9eda-77d5bfcadef0)

NikitaCOEUR commented:
Ah, I can see the whole command in the raw log

neilime commented:
@NikitaCOEUR here is PR that reproduce your usecase (empty `secret-envs` & single `build-secret-github-app-token-env`).

It works fine: https://github.com/hoverkraft-tech/ci-github-container/pull/372

NikitaCOEUR commented:
Quand composer execute git clone, il lance cette commande : 
`git clone --mirror -- https://ghs...V9i:x-oauth-basic@github.com/OWNER/repo.git /tmp/cache/vcs/git-github.com-OWNER-repo.git/`

Whereas, as specified here :  https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app-installation#about-authentication-as-a-github-app-installation
>You can also use an installation access token to authenticate for HTTP-based Git access. Your app must have the "Contents" repository permission. You can then use the installation access token as the HTTP password. Replace TOKEN with the installation access token: git clone https://x-access-token:TOKEN@github.com/owner/repo.git.

And when I try this command by hand, it works fine : 
`
git clone --mirror -- https://x-access-token:ghs_...V9i@github.com/OWNER/repo.git /tmp/cache/vcs/git-github.com-OWNER-repo.git/
`

NikitaCOEUR commented:
Well, with this workaround, it works: 
```
RUN --mount=type=secret,id=GITHUB_TOKEN \
    set -eu -o pipefail; \
    if [ -f /run/secrets/GITHUB_TOKEN ]; then export GITHUB_TOKEN=$(cat /run/secrets/GITHUB_TOKEN); fi; \
    if [ -n "$GITHUB_TOKEN" ]; then \
        if [[ "$GITHUB_TOKEN" == ghs_* ]]; then \
            composer config --global --auth http-basic.github.com x-access-token $GITHUB_TOKEN; \
        else \
            composer config --global github-oauth.github.com $GITHUB_TOKEN; \
        fi; \
    fi;
```

But, inevitably, a problem: a `ghs_*` type token has a limited duration, especially if you don't specify the parameter [`skip-token-revoke`](https://github.com/actions/create-github-app-token?tab=readme-ov-file#skip-token-revoke), by default, it is set to false, so the token generated is revoked at the end of the job. Also, if a second job reuses this cached layer, it will use a revoked token to perform its actions.... So be sure to use the secret mount at the necessary steps, and not in a dedicated layer...

Maybe add a warning in the documentation?

neilime commented:
Pay attention that  `RUN --mount=...` options only persist for the lifetime of their specific RUN command. The expected behavior of `RUN --mount=secret=...` is that the secret is not persisted in the image. So you cannot rely on the secrets in following layers

NikitaCOEUR commented:
Can you prepare a release to take advantage of all these changes?

neilime commented:
> Can you prepare a release to take advantage of all these changes?

@NikitaCOEUR the new version has been released: https://github.com/hoverkraft-tech/ci-github-container/releases/tag/0.21.0
